pkg/util/limiter/utilization.go

```diff
@@ -0,0 +1,330 @@
+	"github.com/grafana/dskit/services"
+	"github.com/pkg/errors"
+	"github.com/prometheus/procfs"
+	"go.uber.org/atomic"
+
+	"github.com/grafana/mimir/pkg/util/math"
+)
+
+const (
+	// Interval for updating resource (CPU/memory) utilization
+	resourceUtilizationUpdateInterval = time.Second
+
+	memPathV2 = "/sys/fs/cgroup/memory.current"
+	cpuPathV2 = "/sys/fs/cgroup/cpu.stat"
+	memPathV1 = "/sys/fs/cgroup/memory/memory.stat"
+	cpuPathV1 = "/sys/fs/cgroup/cpu/cpuacct.usage"
+)
+
+type utilizationScanner interface {
+	// Method returns the method for getting resource utilization.
+	Method() string
+
+	// Scan returns CPU time and memory utilization, or an error.
+	Scan() (float64, uint64, error)
+}
+
+type cgroupV2Scanner struct {
+}
+
+func (s cgroupV2Scanner) Method() string {
+	return "cgroup v2"
+}
+
+func (s cgroupV2Scanner) Scan() (float64, uint64, error) {
+	// For reference, see https://git.kernel.org/pub/scm/linux/kernel/git/tj/cgroup.git/tree/Documentation/admin-guide/cgroup-v2.rst
+	memR, err := readFileNoStat(memPathV2)
+	if err != nil {
+		return 0, 0, err
+	}
+	defer memR.Close()
+	var memUtil uint64
+	if _, err := fmt.Fscanf(memR, "%d", &memUtil); err != nil {
+		return 0, 0, errors.Wrapf(err, "failed scanning %s", memPathV2)
+	}
+
+	// For reference, see https://git.kernel.org/pub/scm/linux/kernel/git/tj/cgroup.git/tree/Documentation/admin-guide/cgroup-v2.rst
+	cpuR, err := readFileNoStat(cpuPathV2)
+	if err != nil {
+		return 0, 0, err
+	}
+	defer cpuR.Close()
+	var name string
+	var cpuTime uint64
+	for {
+		if _, err := fmt.Fscanf(cpuR, "%s %d\n", &name, &cpuTime); err != nil {
+			if errors.Is(err, io.EOF) {
+				break
+			}
+
+			return 0, 0, errors.Wrapf(err, "failed scanning %s", cpuPathV2)
+		}
+
+		if name == "usage_usec" {
+			// This is CPU time in microseconds
+			break
+		}
+	}
+	if name != "usage_usec" {
+		return 0, 0, fmt.Errorf("failed scanning %s", cpuPathV2)
+	}
+
+	return float64(cpuTime) / 1e06, memUtil, nil
+}
+
+func newCgroupV2Scanner() (utilizationScanner, bool) {
+	if _, err := os.Stat(memPathV2); err != nil {
+		return cgroupV2Scanner{}, false
+	}
+	if _, err := os.Stat(cpuPathV2); err != nil {
+		return cgroupV2Scanner{}, false
+	}
+	return cgroupV2Scanner{}, true
+}
+
+type cgroupV1Scanner struct {
+}
+
+func (s cgroupV1Scanner) Method() string {
+	return "cgroup v1"
+}
+
+func (s cgroupV1Scanner) Scan() (float64, uint64, error) {
+	memR, err := readFileNoStat(memPathV1)
+	if err != nil {
+		return 0, 0, err
+	}
+	defer memR.Close()
+	// Summing RSS and cache for memory usage
+	// For reference, see section 5.5 in https://www.kernel.org/doc/Documentation/cgroup-v1/memory.txt
+	var memUtil uint64
+	var numMemStats int
+	for {
+		var name string
+		var val uint64
+		if _, err := fmt.Fscanf(memR, "%s %d\n", &name, &val); err != nil {
+			if errors.Is(err, io.EOF) {
+				break
+			}
+
+			return 0, 0, errors.Wrapf(err, "failed scanning %s", memPathV1)
+		}
+
+		if name == "rss" || name == "cache" {
```

Comment:
Should we make rss and cache constants?

Reply:
Isn't it more hassle than it's worth as these strings are used only once?